Use regex matching to determine icons for SocialLinkAttachements

dulldesk opened the pull request:
Resolves #70 

- Remove `SocialLink` model
- Migration updates current `SocialLinkAttachments` with the domain from `SocialLink` as applicable

dulldesk committed:
Replace SocialLinks with regex matching in SocialLinkAttachements

dulldesk committed:
Create migration

dulldesk committed:
Remove social links fixture

dulldesk commented:
Something to consider, whether now or later:
A link's content can inject JavaScript, e.g. `javascript:alert("hello world")` (on click of the link, this would run) and the like. 
Probably best to clean input when the links are initially created / updated (SocialLinkForm)

fluix-dev commented:
> A link's content can inject JavaScript

This is definitely something that should be address in this PR. A combination of Django's `safe_string` and/or more strict regex matching is the way to go. For the latter portion, I think it's mainly the `Generic URL` that needs safe-url matching.

fluix-dev commented on app/models.py:
There's still a regex for it though, so this should really throw an error.

fluix-dev commented on app/models.py:
```suggestion
        return "fab" if "fab" in self.social_link() else "fas"
```

fluix-dev commented on app/models.py:
This should probably `re.compile` and include the ignore-case flag.

fluix-dev commented on app/models.py:
Just a note that if you ignore case you can then remove the `A-Z` (or `a-z`) from here.

fluix-dev replied:
`mailto:` should be mandatory here.

fluix-dev commented on app/models.py:
`tel:` should be mandatory here.

dulldesk committed:
Update ternary statement

Co-authored-by: Steven Guikal <AvidDev@protonmail.com>

dulldesk committed:
Use re.compile and case-insensitive regex patterns

"tel:" and "mailto:" are also enforced

dulldesk committed:
Escape link content and raise error if link is matchless

dulldesk committed:
Edit ternary statement

dulldesk committed:
Validate and clean link content upon creation

dulldesk committed:
Mark fields as not required

The form automatically "cleans"/fills these fields with the correct values. This is done to prevent errors regarding these fields being missing from popping up upon invalid data in the link content.

dulldesk commented on app/forms.py:
```diff
@@ -31,8 +33,12 @@ def __init__(self, *args, **kwargs):
         self.fields["content_type"].disabled = True
+        self.fields["object_id"].disabled = True
+
+        # These two fields are automatically "cleaned"/filled anyways
+        self.fields["content_type"].required = False
+        self.fields["object_id"].required = False
```
As described in the commit, this is done to prevent errors regarding these fields being missing from popping up upon invalid data in the link content. Is there a better way of accomplishing this (e.g. would filtering out these errors be preferable)?

dulldesk commented on app/forms.py:
```diff
@@ -45,6 +51,19 @@ def clean_content_type(self):
         cleaned_data["content_type"] = ContentType.objects.get_for_model(self.object.__class__)
         return cleaned_data["content_type"]
 
+    def clean_content(self):
+        data = self.cleaned_data["content"]
+        if data.lower().startswith("mailto:"):
+            validate_email(data[7:])
+        elif data.lower().startswith("tel:"):
+            validate_tel = RegexValidator(r"^[0-9]+$", "Enter a valid telephone number")
+            validate_tel(data[4:])
+        else:
+            validate_url = URLValidator()
+            validate_url(data)
+        data = escape(data)
+        return data
+
```
I figured it better to clean the data beforehand, for if a malicious link were to be submitted, the regex would catch it and prevent it from rendering, but would also yield a 500 to that page. Thus, this would be preventative of that situation.

dulldesk committed:
Undo redundant 'fix'

dulldesk committed:
Edit unnecessary assignment

dulldesk committed:
Remove ValueError

dulldesk committed:
Use 'fa' for the key for SOCIAL_LINKS

fluix-dev reviewed (approved):
LGTM, thanks for working on this :smile:
